packages/admin-next/dashboard/src/components/common/infinite-list/infinite-list.tsx

```diff
@@ -0,0 +1,138 @@
+import { QueryKey, useInfiniteQuery } from "@tanstack/react-query"
+import { ReactNode, useEffect, useMemo, useRef } from "react"
+import { Skeleton } from "../skeleton"
+import { toast } from "@medusajs/ui"
+import { Spinner } from "@medusajs/icons"
+import { useTranslation } from "react-i18next"
+
+type InfiniteListProps<TResponse, TEntity, TParams> = {
+  queryKey: QueryKey
+  queryFn: (params: TParams) => Promise<TResponse>
+  queryOptions?: { enabled?: boolean }
+  renderItem: (item: TEntity) => ReactNode
+  responseKey: keyof TResponse
+  pageSize?: number
+}
+
+export const InfiniteList = <
+  TResponse extends { count: number; offset: number; limit: number },
+  TEntity extends { id: string },
+  TParams extends { offset?: number; limit?: number },
+>({
+  queryKey,
+  queryFn,
+  queryOptions,
+  renderItem,
+  responseKey,
+  pageSize = 20,
+}: InfiniteListProps<TResponse, TEntity, TParams>) => {
+  const { t } = useTranslation()
+
+  const {
+    data,
+    error,
+    fetchNextPage,
+    fetchPreviousPage,
+    hasPreviousPage,
+    hasNextPage,
+    isFetching,
+    isPending,
+  } = useInfiniteQuery({
+    queryKey: queryKey,
+    queryFn: async ({ pageParam = 0 }) => {
+      return await queryFn({
+        limit: pageSize,
+        offset: pageParam,
+      } as TParams)
+    },
+    initialPageParam: 0,
+    maxPages: 5,
+    getNextPageParam: (lastPage) => {
+      const moreItemsExist = lastPage.count > lastPage.offset + lastPage.limit
+      return moreItemsExist ? lastPage.offset + lastPage.limit : undefined
+    },
+    getPreviousPageParam: (firstPage) => {
+      const moreItemsExist = firstPage.offset !== 0
+      console.log(
+        firstPage,
+        moreItemsExist
+          ? Math.max(firstPage.offset - firstPage.limit, 0)
+          : undefined
+      )
+
+      return moreItemsExist
+        ? Math.max(firstPage.offset - firstPage.limit, 0)
+        : undefined
+    },
+    ...queryOptions,
+  })
+
+  const items = useMemo(() => {
+    return data?.pages.flatMap((p) => p[responseKey] as TEntity[]) ?? []
+  }, [data, responseKey])
+
+  const parentRef = useRef<HTMLDivElement>(null)
+  const startObserver = useRef<IntersectionObserver>()
+  const endObserver = useRef<IntersectionObserver>()
+
+  useEffect(() => {
+    if (isPending) {
+      return
+    }
+
+    // Clear the old observers
+    startObserver.current?.disconnect()
+    endObserver.current?.disconnect()
```

Comment:
suggestion: we should do this in a function that is returned from the effect as per react convention

Reply:
This particular disconnect has to happen before we create the new ones, but I added a new empty useEffect to clean up on unmount. WDYT?

Reply:
But the new observer isn't necessary, all we need is to move return function from your new effect to this effect and everything will work as we expect.

On the initial pass, there are no observers so we don't need cleanup before. After first, pass a cleanup function is registered and React will call it before it runs the effect for the second time and so on.